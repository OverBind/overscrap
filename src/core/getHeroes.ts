import puppeteer from 'puppeteer';
import { Hero } from '../types';
import { HeroID, HeroStat } from '../enums';

export async function getHeroes(page: puppeteer.Page): Promise<Hero[]> {
  return page.evaluate(
    (HeroID, HeroStat) => {
      const list: Hero[] = [];
      const items = document.querySelectorAll(
        'div#competitive select.dropdown-select-element option'
      );

      items.forEach((item) => {
        const id: string = item.getAttribute('value') || '';
        const name = item.innerHTML;
        const isHero = Object.values(HeroID).find(
          (heroID) => heroID === id && id !== HeroID.ALL_HEROES
        );
        if (isHero) {
          const hero = {
            id,
            name,
            src: `https://d1u1mce87gyfbn.cloudfront.net/game/heroes/small/${id}.png`,
            info: Object.values(HeroStat).map(
              (dataId) => {
                const getHeroStats: HTMLInputElement | null =
                  document.querySelector(
                    `div#competitive div[data-category-id = '${id}']`
                  );
                if (getHeroStats) {
                  const data = Array.from(
                    getHeroStats.querySelectorAll(
                      `tr[data-stat-id = '${dataId}'] td`
                    ),
                    (col) => col.innerHTML
                  );
                  return {
                    name: data[0],
                    value: data[1]
                  };
                }
                return {
                  name: '',
                  value: ''
                };
              }
            )
          };
          list.push(hero);
        }
      });
      return list;
    },
    HeroID,
    HeroStat
  );
}
