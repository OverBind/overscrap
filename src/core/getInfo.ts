import puppeteer from 'puppeteer';
import { GameStat } from '../enums';
import { InfoField } from '../types';

export async function getInfo(page: puppeteer.Page): Promise<InfoField[]> {
  return page.evaluate((GameStat) => {
    return Object.values(GameStat).map((dataId) => {
      const row = Array.from(
        document.querySelectorAll(`tr[data-stat-id = '${dataId}']`)
      );
      const data = Array.from(row[1].children, (td) => td.innerHTML);
      return {
        name: data[0],
        value: data[1]
      };
    });
  }, GameStat);
}
