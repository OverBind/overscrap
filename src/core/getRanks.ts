import puppeteer from 'puppeteer';
import { Rank } from '../types';

export async function getRanks(page: puppeteer.Page): Promise<Rank[]> {
  return page.evaluate(() => {
    return Array.from(
      document.querySelectorAll(
        'div.masthead-player div.competitive-rank-role'
      ),
      (rank) => ({
        src: rank.querySelector<HTMLInputElement>(
          'img.competitive-rank-tier-icon'
        )?.src,
        role: rank
          .querySelector<HTMLInputElement>('span.ow-Tooltip')
          ?.innerText.replace(/ .*/, ''),
        sr: rank.querySelector<HTMLInputElement>('div.competitive-rank-level')
          ?.innerText
      })
    );
  });
}
