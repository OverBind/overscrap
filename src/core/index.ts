export { getHeroes } from './getHeroes';
export { getInfo } from './getInfo';
export { getRanks } from './getRanks';