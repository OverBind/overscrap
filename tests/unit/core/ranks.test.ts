/**
 * @jest-environment jsdom
 */
 import { setupTest } from '../../helper/setupTest';
 import { getRanks } from '../../../src/core/getRanks';
 
 describe('Get player stat', () => {
   test('Should get player stat', async () => {
     const page = await setupTest();
 
     const data = await getRanks(page);
 
     expect(data).toBeTruthy();
     expect(data).toMatchSnapshot();
   });
 });
 