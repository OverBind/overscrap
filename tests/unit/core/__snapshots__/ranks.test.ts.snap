// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Get player stat Should get player stat 1`] = `
Array [
  Object {
    "role": "Tank",
    "sr": "2918",
    "src": "https://d1u1mce87gyfbn.cloudfront.net/game/rank-icons/rank-PlatinumTier.png",
  },
]
`;
